### ISSUE TITLE
Suggestion: Less picky quarantine

### ISSUE THREAD

**Issue by Foxandxss** [DevExpress/testcafe#5334]
### What is your Test Scenario?
I am testing a large Angular application and there are some flakes. Testcafé does its best to prevent them but it is not always possible. Sometimes you have a test like:

```typescript
await t
  .click(foo)
  .expect(somethingHappenedAfterClick);
```

Nothing too special, but sometimes for reasons I have yet to understand, Angular receives the click but decides not to do anything with that, just swallow it. I would have to click it again to make the expectation.

In short, a flaky test

### What are you suggesting?
I discovered the quarantine mode today and it is INCREDIBLE. I have many many many tests and some of them fails (I have them identified).

From 10-12 flaky tests, with quarantine I am down to 2 unstable.

Are they really unstable? Not in terms of Fowler, it is a different thing in my opinion.

I have for example a dropdown with a sign out button and sometimes clicking the dropdown shows it and sometime doesn't. Maybe in 5 runs, fails 2. But the important thing here is that they only fail within testcafé. In the real app it works 100/100.

I would love for quarantine mode to have an option for unstable tests to not mark them as failures. Just a warning that says "Hey, this test need attention" but not to fail the test itself since it did pass at least once".

I understand that this is not its purpose, but it will make the life way way better for flaky tests that are outside your reach to fix.

### What alternatives have you considered?
I considered other frameworks to see if they weren't having this issue. I have many tests (in different applications) that for some reason swallow clicks. It is like testcafé trying to do its job before Angular finishes rendering. I know there is an angular plugin but I didn't see a difference.

### Additional context
Maybe this is not the PROPER solution for this, maybe there is a bug in testcafé with Angular, but in the meantime, this quarantine made my day.

**Comment by alexey-lin** [comment]
Hi @Foxandxss,

Thank you for the suggestion, we will discuss it with the team. At the moment, a flaky test counts as failed if it fails more often than it passes (e.g. 3 times of 5 attempts).

As for the plugin, yes, we have the [`testcafe-angular-selectors` plugin](https://github.com/DevExpress/testcafe-angular-selectors). Did you try calling the `waitForAngular` function before every test (please see the [Examples](https://github.com/DevExpress/testcafe-angular-selectors#examples) section of the plugin's README)? This should prevent TestCafe from interacting with a page before Angular rendering is completed.

**Comment by Foxandxss** [comment]
Hello @alexey-lin 

I would like to group E2E tests like this:

a) Tests that always work as intended
b) Tests that normally pass but depends on something third party that could make them fail
c) Automated tool not behaving like it should
d) CI machine is not as fast as yours.

*a* and *d* are the easier ones. You can always tweak your tests so it is not speed dependant. You can always bump some timeout and you're good to go.

Quarantine is very nice with *b* because you will normally get a green result and if you start getting too many *unstable* then you know you need to work in it.

The problem lies with *c*. I have been doing E2E for over 5 years as a full time job but only with Angular and I have seen some strange patterns in different projects and I am yet unable to pinpoint a cause.

I have for example a page where after clicking a button and doing a process, it automatically redirects to a new one. For unknown reasons while executing this process within TestCafé the new page gets half loaded and I have to trigger a `.click` so it keeps loading. That is an easy solution, never flakes.

Then I have a case where I have a custom radio button that gets an `active` css class when selected. I have a test where I assert that this class applied (and that the other button would get it removed) and also some validation errors would appear. For unknown reasons there are some times where the `active` css class is not removed when toggling the radio buttons. Ok, here you can say that my code is wrong somehow, but you can toggle them 1.000 times on the real app and it would never ever do that but then in testcafé I can see visually how the two buttons gets the same CSS class when toggling.

**It is like Angular behaves differently when Testcafé is driving it**.

So this test kind of flakes on me.

Then I have many many cases (in separate projects even with different UI libraries and stuff) where I click a link or button that would trigger something (new page, section appears, etc) and Angular will just swallow the click and do nothing. Again this would *never ever* happen when using the app for real, but it does happen inside testcafé.

This tests flakes on mes as well.

I did try `testcafe-angular-selectors` but I didn't get anything different so I guess the problem comes from somewhere else.

I don't know if Angular is doing it wrong or Testcafé has a strange bug when using Angular.

Now is when we come to Quarantine. I can say that 90% of this flaky tests will happen once or twice every 10 executions so with quarantine mode I get way better results.

Am I misusing quarantine? I may, but when you get say 10 out of 200 tests that may fail or not and there is nothing you can do to fix them, it gets problematic.

My solution was to skip them with a comment saying "This is a flaky test". That is bad. I am losing value in there.

Now we get to the 10% that even quarantine doesn't set them straight. Right now I am putting the "This is a flaky test" comment on them. Better to get it on 1 every 200 tests. Those are still tests out of my control, that testcafé is making my application behave differently and sometimes I can hack through it and sometimes I can't.

So I would like a way to know when a test is unstable so I can do something about it or even a way to say "hey, run it up to 3-5 times, if it pass at least once, just tell me it wasn't that stable but move on".

To finish, let me share a hack I use. After creating an account I am presented with two choices, when I click one (it is a button) I get some extra options to continue. Sometimes when testcafé clicks the button it does nothing, even a second click does nothing and then maybe the third click goes in and everything moves forward. Since I do this process many times, I had to write this:

```
await t.click(sut.firstOption);


 for (let i = 0; i < 5; i++) {
    // Click didn't go through?
    const isVisible = await sut.wizard.visible;

    if (!isVisible) {
      await t.wait(1000).click(sut.firstOption);
      continue;
    }

    break;
  }
```

Basically it says click the button. If what should appear after it is not there (click was swallowed) try it again up to 5 times.

Sorry for the long answer, but I believe it was worth the insight.

**Comment by AlexKamaev** [comment]
@Foxandxss,

Thank you for your detailed explanation. As far as I understand, you have issues with event processing with Angular + TestCafe, and so your tests are flaky ( i.e. Angular "swallows" click? radio button does not get "active" class and so on).
In this case I think it's better to fix the cause of the issue but not to hide it with quarantine mode. Please share your project with us (or create a sample project) and we will try to find the cause of the issue and fix it if it's on our side.

**Comment by Foxandxss** [comment]
I will try to reproduce it in a new project since this is a client work. But yes, I am with you here, if this is an issue with testcafe, we should be able to fix it.

**Comment by Ogurecher** [comment]
@Foxandxss, Thank you for your collaboration. Feel free to create a separate issue if you reproduce this behavior.
